File: src/CSharp/601-700/606.Tree2str.cs

using System;
using System.Collections.Generic;
using System.Text;

namespace PeefyLeetCode.Tree2str
{
    public class TreeNode
    {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode(int x) { val = x; }
    }

    public class Solution
    {
        public string Tree2str(TreeNode t)
        {
            StringBuilder result = new StringBuilder();
            tree2str(t, result);

            return result.ToString();
        }
        void tree2str(TreeNode t, StringBuilder result)
        {

            if (t == null)
            {
                return;
            }

            result.Append(t.val);

            if (t.left != null)
            {
                result.Append("(");
                tree2str(t.left, result);
                result.Append(")");
            }

            if (t.right != null)
            {

                if (t.left == null)
                {
                    result.Append("()(");
                }
                else
                {
                    result.Append("(");
                }

                tree2str(t.right, result);
                result.Append(")");
            }

        }
    }
}

File: src/Python/601-700/606.Tree2str.py

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """ 
        if not t:
            return ''
        s = str(t.val)
        left = self.tree2str(t.left)
        right = self.tree2str(t.right)
        if not left and not right:
            return s
        if not right:
            return s+'('+left+')'
        return s+'('+left+')'+'('+right+')'

if __name__ == '__main__':
    solution = Solution()
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    print(solution.tree2str(root))
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(4)
    print(solution.tree2str(root))
else:
    pass
